### PR TITLE
[FIX] website: remove ToC snippet on removing last content 

### PR DESCRIPTION
**Pull request by can-odoo**
The user has the option to remove the last block from the Table of
Content snippet prior to this commit, but doing so breaks the Add Item
feature because there is no reference to adding a new item in the
snippet area, and doing so is also not a good idea.

If the block is the final one in the Table of Content after this commit,
removing the last Content will remove the ToC snippet.

---

Before this PR, when we remove the ToC snippet, underneath the hood
navbar position of the ToC snippet was still trying to update which
should not. This is because we have bound the
`_updateTableOfContentNavbarPosition` method to scrolling events so
whenever we scroll up/down navbar position is trying to update, which is
the intended behavior when we have a ToC snippet available, but
unexpected behavior when the snippet is removed.
With this fix, we prevent that by unbound the method when the public
widget is destroyed.

---

task-3081442

### CODE REVIEW

**Comment by can-odoo** [comment]
Hello @rdeodoo ,
can you please review the code 👨🏽‍💻 ❓

File: addons/website/static/src/snippets/s_table_of_content/options.js [hunk 1]
```diff
@@ -56,7 +56,10 @@ options.registry.TableOfContent = options.Class.extend({
             $el.attr('id', id);
             $el[0].dataset.anchor = 'true';
         });
-        $nav.find('a:first').addClass('active');
+        const targetNode = this.$target.find('.s_table_of_content_main')[0]
```
Comment (Guillaume-gdi) [inline addons/website/static/src/snippets/s_table_of_content/options.js:59]: - Missing semicolon at the end of the line
- targetNode is a HTMLElement so the variable name should finish with `El`
- Maybe you can avoid adding new JQuery code (as we want to get rid of it) => `this.$target[0].querySelector(...`

File: addons/website/static/src/snippets/s_table_of_content/options.js [hunk 2]
```diff
@@ -56,7 +56,10 @@ options.registry.TableOfContent = options.Class.extend({
...
+        targetNode.children.length === 0
+            ? (this.trigger_up('go_to_parent', {$snippet: this.$target}), this.$target.remove())
+            : $nav.find('a:first').addClass('active');
```
Comment (Guillaume-gdi) [inline addons/website/static/src/snippets/s_table_of_content/options.js:62]: When it's not an assignation, we encourage if else instead of ternary operator :+1:

**Review by Guillaume-gdi — CHANGES_REQUESTED** [review]
Hello, here is a first review.

I wonder why we don't use `content_changed` to remove the TOC when there is no more child. (I guess that `will_remove_snippet` doesn't work...)

I also notice that when we remove last content with backspace we have a traceback :grimacing: Can you check if we can provide a fix for that too ?

Finally, don't hesitate to check the [commit message rules](https://www.odoo.com/documentation/16.0/contributing/development/git_guidelines.html#commit-message-structure). Check your spelling `the block or content` => `the block of content`. Max 72 char per row.

**Comment by Guillaume-gdi** [comment]
@can-odoo You can also remove the PR link in you PR description. As we are already on it, it's useless :+1: 
image

**Comment by rdeodoo** [comment]
@robodoo delegate=Guillaume-gdi

**Comment by can-odoo** [comment]
> I wonder why we don't use `content_changed` to remove the TOC when there is no more child.

I have tried using `content_changed` like below but it's not working as expected
```js
this.$target.on('content_changed', function(event) {
    if (targetNodeEl && targetNodeEl.children.length === 0) {
        targetNodeEl.offsetParent.parentElement.remove()
    }
});
```

> I also notice that when we remove last content with backspace we have a traceback

🤦🏽 checking this one, it seems from `pluginEvents` of `summernote`

**Comment by can-odoo** [comment]
Hello @Guillaume-gdi,
Changes have been applied.

File: addons/website/static/src/snippets/s_table_of_content/options.js [hunk 3]
```diff
@@ -56,7 +56,15 @@ options.registry.TableOfContent = options.Class.extend({
             $el.attr('id', id);
             $el[0].dataset.anchor = 'true';
         });
-        $nav.find('a:first').addClass('active');
+
+        const targetNodeEl = this.$target[0].querySelector('.s_table_of_content_main');
+        if (targetNodeEl.children.length === 0) {
+            this.trigger_up('go_to_parent', {$snippet: this.$target});
+            // Remove the ToC since there are no more child elements
```
Comment (Guillaume-gdi) [inline addons/website/static/src/snippets/s_table_of_content/options.js:63]: Put a point at the end of your sentences :+1:

File: addons/web_editor/static/src/js/editor/summernote.js [hunk 4]
```diff
@@ -1599,7 +1599,12 @@ $.summernote.pluginEvents.backspace = function (event, editor, layoutInfo) {
             before = false;
             prev = dom.hasContentAfter(dom.ancestorHaveNextSibling(node));
         }
-        dom.removeSpace(prev.parentNode, prev, 0, prev, 0); // clean before jump for not select invisible space between 2 tag
+        if (prev && prev.parentNode) {
+            dom.removeSpace(prev.parentNode, prev, 0, prev, 0); // clean before jump for not select invisible space between 2 tag
+        } else {
+            event.preventDefault();
+            return false;
+        }
```
[the remark this answers is not in the document]
Comment (can-odoo) [reply addons/web_editor/static/src/js/editor/summernote.js:1607]: @Guillaume-gdi 
> Explain why Toc is a special case

Not a special case, you can regenerate the issue for other snippets as well. For now, I have mentioned the `Team snippet` also on how to reproduce the issue.

Also just want to mention here, we are no longer using the `summernote` lib from v15.0, so need to fix this only for v14.0.

[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/web_editor/static/src/js/editor/summernote.js:1607]: If this a summernote bug only in 14.0 (which are always the worst possible ones to fix), a possibility is to ignore it as that version will be deprecated and not supported in 3-4 months.
That is, of course, if the bug is not too critical and impacting too many people, which is probably not the case here.

File: addons/website/static/src/snippets/s_table_of_content/options.js [hunk 5]
```diff
@@ -56,7 +56,15 @@ options.registry.TableOfContent = options.Class.extend({
             $el.attr('id', id);
             $el[0].dataset.anchor = 'true';
         });
-        $nav.find('a:first').addClass('active');
+
+        const targetNodeEl = this.$target[0].querySelector('.s_table_of_content_main');
```
[the remark this answers is not in the document]
Comment (can-odoo) [reply addons/website/static/src/snippets/s_table_of_content/options.js:60]: > too bad there's no better place to do it than in this function 

Can't get your point here, do you mean outside of the `_generateNav` function?

[the remark this answers is not in the document]
Comment (Guillaume-gdi) [reply addons/website/static/src/snippets/s_table_of_content/options.js:60]: yes but as you said:
> I have tried using content_changed like below but it's not working as expected

[the remark this answers is not in the document]
Comment (can-odoo) [reply addons/website/static/src/snippets/s_table_of_content/options.js:60]: yeah, it's not working as expected using `content_changed`. What do you suggest?

Apart from that, changes have been made, what do you think?

**Review by Guillaume-gdi — CHANGES_REQUESTED** [review]
I think there is an issue with actual solution. In edit mode
- I drop a ToC block

=> The options and the public widget start. Great
- I remove the Toc block via the remove button of the block

=> The options and the public widget are destroyed. Great
- I drop a ToC block
- I remove all it's inner block one by one until the Toc is removed

=> The options is destroyed. Great but the public widget is not. That is wrong. The public widget should be also destroyed in this case.

Can you check and adapt please. Thank you!

**Comment by can-odoo** [comment]
> The options is destroyed. Great but the public widget is not. That is wrong. The public widget should be also destroyed in this case.

This seems not related to my fix (bug 14.0 to master), the public widget is not destroyed when we remove blocks one by one because [_stopWidgets](https://github.com/odoo/odoo/blob/14.0/addons/web/static/src/js/public/public_root.js#L197) is return `null` and not destroy widget. I have not checked into much but while removing a particular block we get `$from: section.pb16.o_colored_level` and for that condition is not satisfied to destroy the widget.
While we remove the whole ToC snippet, conditions are satisfied and we call `widget.destroy`, Public widget destroyed.

File: addons/website/static/src/snippets/s_table_of_content/options.js [hunk 6]
```diff
@@ -56,7 +56,15 @@ options.registry.TableOfContent = options.Class.extend({
...
+        if (targetNodeEl && targetNodeEl.children.length === 0) {
```
[the remark this answers is not in the document]
Comment (can-odoo) [reply addons/website/static/src/snippets/s_table_of_content/options.js:61]: >Why do you need to test if targetNodeEl exists ?

for v15 (removing content with backspace) it will lead to removing the `targetNodeEl`, so in the end, we do not have `targetNodeEl` and that leads us to TB `targetNodeEl is null` if we are not testing for `targetNodeEl`

**Review by Guillaume-gdi — CHANGES_REQUESTED** [review]
> This seems not related to my fix (bug 14.0 to master), the public widget is not destroyed when we remove blocks one by one because [_stopWidgets](https://github.com/odoo/odoo/blob/14.0/addons/web/static/src/js/public/public_root.js#L197) is return null and not destroy widget. I have not checked into much but while removing a particular block we get $from: section.pb16.o_colored_level and for that condition is not satisfied to destroy the widget.
While we remove the whole ToC snippet, conditions are satisfied and we call widget.destroy, Public widget destroyed.

@can-odoo Maybe it's not related to your changes here but it has to be fixed. We should no endup with a public widget running but it's related block has been deleted.
For example:
- Drop a ToC
- Remove every inner blocks
- Scroll a up and down

=> The public widget function `_updateTableOfContentNavbarPosition` is called and could caused issues. You should destroy the public widget when you remove the block.

**Comment by can-odoo** [comment]
if we focus on `_updateTableOfContentNavbarPosition` this function, it is called not only while removing `every inner blocks` separately and scrolling, but even if we try
> `I remove the Toc block via the remove button of the block`,

The options and the public widget are destroyed, `agreed` but if we still try to scroll up and down, first `destroy` and `_updateTableOfContentNavbarPosition` gets called.

**Review by Guillaume-gdi — CHANGES_REQUESTED** [review]
> The options and the public widget are destroyed, agreed but if we still try to scroll up and down, first destroy and _updateTableOfContentNavbarPosition gets called.

Ok that is a bug too. Why should we update the ToC navbar position if the ToC has been deleted? Ideally this PR should fix it too then. And I maintain that the public widget should be destroyed when the ToC is removed.

**Comment by can-odoo** [comment]
@Guillaume-gdi
Can you please check? I have unbound the `_updateTableOfContentNavbarPosition` after the public widget is destroyed.

File: addons/website/static/src/snippets/s_table_of_content/000.js [hunk 7]
```diff
@@ -16,12 +16,22 @@ const TableOfContent = publicWidget.Widget.extend({
         this.$scrollingElement = this.$target.closest(".s_table_of_content").closestScrollable();
         this.previousPosition = -1;
         this._updateTableOfContentNavbarPosition();
-        extraMenuUpdateCallbacks.push(this._updateTableOfContentNavbarPosition.bind(this));
+
+        //  bind _updateTableOfContentNavbarPosition.
```
Comment (Guillaume-gdi) [inline addons/website/static/src/snippets/s_table_of_content/000.js:20]: Don't add double space please.

File: addons/website/static/src/snippets/s_table_of_content/000.js [hunk 8]
```diff
@@ -16,12 +16,22 @@ const TableOfContent = publicWidget.Widget.extend({
...
     },
     /**
      * @override
      */
     destroy() {
+        //  remove the bound function from the extraMenuUpdateCallbacks array so that
+        //  `_updateTableOfContentNavbarPosition` won't gets called after public widget
+        //  destroyed.
+        const indexOfCallback = extraMenuUpdateCallbacks.indexOf(this.boundUpdateNavbar);
+        if (indexOfCallback !== -1) {
```
Comment (Guillaume-gdi) [inline addons/website/static/src/snippets/s_table_of_content/000.js:32]: How can we end up with `extraMenuUpdateCallbacks` that does not contain `this.boundUpdateNavbar` ? :thinking:

**Review by Guillaume-gdi — CHANGES_REQUESTED** [review]
The public widget on the ToC is still not destroyed after the remove of the last ToC inner content... :/

**Comment by can-odoo** [comment]
> The public widget on the ToC is still not destroyed after the remove of the last ToC inner content... :/

okay, I thought you were `And I maintain that the public widget should be destroyed when the ToC is removed.`

File: addons/website/static/src/snippets/s_table_of_content/options.js [hunk 9]
```diff
@@ -56,7 +56,17 @@ options.registry.TableOfContent = options.Class.extend({
...
             $el[0].dataset.anchor = 'true';
         });
-        $nav.find('a:first').addClass('active');
+
+        const tocMainEl = this.$target[0].querySelector('.s_table_of_content_main');
+        if (tocMainEl && tocMainEl.children.length === 0) {
+            this.trigger_up('go_to_parent', {$snippet: this.$target});
+            // destroy public widget and remove the ToC since there are no more
+            // child elements.
+            this.trigger_up('will_remove_snippet', {$target: this.$target});
```
Comment (can-odoo) [inline addons/website/static/src/snippets/s_table_of_content/options.js:65]: @Guillaume-gdi
not sure but I have tried this way so that removing the last inner ToC will destroy the public widget, let me know your thoughts.

File: addons/website/static/src/snippets/s_table_of_content/000.js [hunk 10]
```diff
@@ -16,12 +16,20 @@ const TableOfContent = publicWidget.Widget.extend({
...
      */
     destroy() {
+        // remove the bound function from the extraMenuUpdateCallbacks array so
+        // that `_updateTableOfContentNavbarPosition` won't gets called after
+        // public widget destroyed.
+        const indexOfCallback = extraMenuUpdateCallbacks.indexOf(this.boundUpdateNavbar);
+        extraMenuUpdateCallbacks.splice(indexOfCallback, 1);
         this.$target.css('top', '');
         this.$target.find('.s_table_of_content_navbar').css({top: '', maxHeight: ''});
         this._super(...arguments);
```
Comment (Guillaume-gdi) [inline addons/website/static/src/snippets/s_table_of_content/000.js:35]: Can you put that change in a separate commit. It's not about "remove ToC snippet on removing last content"
thanks :+1:

**Review by Guillaume-gdi — CHANGES_REQUESTED** [review]
The code is OK, thanks for the changes. Just a small change, then I will merge it :rocket:

**Comment by Guillaume-gdi** [comment]
Hello @can-odoo I see that there is an issue with [enterprise runbot test](https://runbot.odoo.com/runbot/batch/1125137/build/48787317). You probably just need to rebase this PR. Can you check ?

**Review by Guillaume-gdi — APPROVED** [review]
Thanks for the PR @can-odoo.
As you can see [here](https://github.com/odoo/odoo/compare/0ec6c2551eb48dac7f6b9c35ac4026a08b3bfc24..c8a6c3845678a192a1a992afae5dcb6ab554049f), I just removed those comments because they were not very useful IMO and I updated the task description.

@robodoo rebase-ff r+
